### PR TITLE
New cni selection using cluster class variable

### LINKED ISSUE #1617: CNI selection using cluster class variable
CNI selection should happen in tanzuclusterbootstrap_controller.go using the following

1. Update `TanzuClusterBootstrapTemplate` spec.cni to spec.cnis to indicate multiple supported cnis
2. Controller should pick the cni that matches a cluster class variable or default to the first one in the template spec.cnis

### PR DESCRIPTION
**Pull request by maralavi**
### What this PR does / why we need it
This PR makes cni selection using cluster class variable. If not cluster variable is defined, it uses the first entry defined in the template as the cnu.

### Which issue(s) this PR fixes

Fixes https://github.com/vmware-tanzu/tanzu-framework/issues/1617

### Describe testing done for PR

Added unit tests

### Release note

```release-note
package-based-lcm: make cni selection using cluster class variable
```

### PR Checklist

- [X] Squash the commits into one or a small number of logical commits
      
- [X] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [X] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

File: addons/controllers/tanzuclusterbootstrap_controller.go
Comment (vijaykatam) [inline addons/controllers/tanzuclusterbootstrap_controller.go:660]: Should process only type `string`, for others return error.

File: addons/controllers/tanzuclusterbootstrap_controller.go
Comment (vijaykatam) [inline addons/controllers/tanzuclusterbootstrap_controller.go:368]: cni selection should happen here

File: addons/controllers/tanzuclusterbootstrap_controller.go
Comment (vijaykatam) [inline addons/controllers/tanzuclusterbootstrap_controller.go:602]: I would suggest something below that gets called after `tanzuClusterBootstrap.Spec = tanzuClusterBootstrapTemplate.Spec.DeepCopy()` instead of patching the object after it is created.
```
func (r *TanzuClusterBootstrapReconciler) getCNI(tanzuClusterBootstrapTemplate *runtanzuv1alpha3.TanzuClusterBootstrapTemplate, cluster *clusterapiv1beta1.Cluster, log logr.Logger) (*runtanzuv1alpha3.TanzuClusterBootstrapPackage, error)
```
Once CNIs has been set and tanzuClusterBootstrap object created CNIs is immutable and will be enforced by a webhook.

File: addons/controllers/tanzuclusterbootstrap_controller.go
Comment (maralavi) [reply addons/controllers/tanzuclusterbootstrap_controller.go:369]: I changed to do right before this function ```r.cloneSecretsAndProviders``` as ```createOrPatchTanzuClusterBootstrapFromTemplate``` seemed a better place to add and didn't need passing template object to ```cloneSecretsAndProviders```

File: addons/main.go
Comment (blc1996) [inline addons/main.go:155]: Do we want to add ClusterBootstrap controller to main.go now?

File: addons/main.go
Comment (blc1996) [inline addons/main.go:158]: Can we create a config struct like `addonconfig.Config`? Some more configs will be added.

File: addons/main.go
Comment (maralavi) [reply addons/main.go]: Not yet, just removed.

File: addons/main.go
Comment (maralavi) [reply addons/main.go]: Sure, done

File: addons/controllers/testdata/test-cluster-bootstrap.yaml
Comment (blc1996) [inline addons/controllers/testdata/test-cluster-bootstrap.yaml:35]: Look like the CLI team hasn't finalized the naming convention for clusterClass vars.
Saw something in Anuj's [PR](https://github.com/vmware-tanzu/tanzu-framework/pull/1668/files#diff-64eae10d16908dfcb6b093a12243425fbb7b16e7c8c0d8573a883b9ea39a371cR41)

File: addons/controllers/clusterbootstrap_controller.go
Comment (blc1996) [inline addons/controllers/clusterbootstrap_controller.go:658]: Can we make this a generic function that can be used to fetch different clusterClass vars?

Comment (maralavi) [reply addons/controllers/clusterbootstrap_controller.go:658]: I don't think so. the reason is that each variable might have a different type in the schema (string, numeric, slice, ...) and this function would get too hacky if made generic to accommodate for all possible variable types. I'd rather having separate functions for different variables for now.

Comment (blc1996) [reply addons/controllers/clusterbootstrap_controller.go:658]: Those we need will all be strings. Basically what's in https://github.com/vmware-tanzu/tanzu-framework/blob/main/pkg/v1/providers/config_default.yaml

Comment (blc1996) [reply addons/controllers/clusterbootstrap_controller.go:658]: Maybe a generic function like `GetClusterVariableStringByName` for string would be enough.

Comment (maralavi) [reply addons/controllers/clusterbootstrap_controller.go:658]: Yes I can do one only for string type variables

Comment (maralavi) [reply addons/controllers/clusterbootstrap_controller.go]: Added ```parseClusterVariableString()``` for parsing string variables

File: addons/controllers/clusterbootstrap_controller.go
Comment (vijaykatam) [inline addons/controllers/clusterbootstrap_controller.go:373]: This line could be moved up into the previous append

File: addons/controllers/testdata/test-cluster-bootstrap.yaml
Comment (vijaykatam) [reply addons/controllers/testdata/test-cluster-bootstrap.yaml:35]: Should be fine as long as the name is configurable. I think the var names will differ between TKGm and TKGS due to TKGm needing to support legacy configs.

File: addons/controllers/clusterbootstrap_controller.go
Comment (vijaykatam) [inline addons/controllers/clusterbootstrap_controller.go:658]: Could this be a standalone helper in pkg/util? Nothing specific to ClusterBootstrapReconciler in the method. 

Suggestion: Move to pkg/util and write a couple of unit tests.

File: addons/controllers/clusterbootstrap_controller.go
Comment (vijaykatam) [inline addons/controllers/clusterbootstrap_controller.go:665]: nitpick: error message should "cluster is nil"

File: addons/controllers/clusterbootstrap_controller.go
Comment (vijaykatam) [inline addons/controllers/clusterbootstrap_controller.go:667]: or if `variableName` is empty

File: addons/controllers/testdata/test-cluster-bootstrap.yaml
Comment (vijaykatam) [inline addons/controllers/testdata/test-cluster-bootstrap.yaml:37]: Could we add a test for the case when cluster variable does not exist?

File: addons/controllers/clusterbootstrap_controller.go
Comment (maralavi) [reply addons/controllers/clusterbootstrap_controller.go:372]: No, cause now both ```CNIs``` and ```AdditionalPackages``` are slices, it doesn't let me to do that.

**Review by vijaykatam — COMMENTED** [review]
Changes LGTM. Will approve once capi bump is merged and tests pass.

**Review by vijaykatam — APPROVED** [review]
LGTM. Good test coverage!

**Pull request merged**
